sway/input/cursor.c

```diff
@@ -539,10 +457,14 @@ static void handle_touch_motion(struct wl_listener *listener, void *data) {
 		seat->touch_x = lx;
 		seat->touch_y = ly;
 
-		struct sway_drag_icon *drag_icon;
-		wl_list_for_each(drag_icon, &root->drag_icons, link) {
-			if (drag_icon->seat == seat) {
-				drag_icon_update_position(drag_icon);
+		struct wlr_scene_node *node;
+		wl_list_for_each(node, &seat->scene_node->state.children, state.link) {
+			struct sway_scene_descriptor *desc = node->data;
+
+			if (desc) {
```

Comment:
Can `desc` ever be null? (the previous code didn't null-check it)

Reply:
Yeah, can be null if the scene descriptor allocation failed. Where is the place you're finding I'm not null checking?

Reply:
I meant that prior to your force-push this code read `if (desc->type == SWAY_SCENE_DESC_DRAG_ICON)`, which implied that `desc != NULL`. I would prefer rejiggering the allocation code a bit so that the node just isn't created if the scene descriptor fails to allocate. There's no real reason to try to chug along in a half-broken after an allocation fails, things are already on fire -- unless we can bubble up the allocation failure such that the action that triggered it becomes a no-op. (Not always possible, usually impossible to actually test.)

The fewer non-semantic null-checks there are, the easier it is to reason about C code.